gcc/rust/ast/rust-ast-dump.cc

```diff
@@ -62,28 +57,86 @@ Dump::go (AST::Item &item)
   item.accept_vis (*this);
 }
 
+template <typename T>
 void
-Dump::format_function_param (FunctionParam &param)
+Dump::visit (T *node)
 {
-  param.get_pattern ()->accept_vis (*this);
-  stream << ": ";
-  param.get_type ()->accept_vis (*this);
+  node->accept_vis (*this);
 }
 
+template <typename T>
 void
-Dump::emit_attrib (const Attribute &attrib)
+Dump::visit (std::unique_ptr<T> &node)
 {
-  stream << "#[";
+  node->accept_vis (*this);
+}
+
+template <typename T>
+void
+Dump::visit_items_joined_by_separator (T &collection,
+				       const std::string &separator)
+{
+  if (collection.size () > 0)
+    {
+      visit (collection.at (0));
+      for (size_t i = 1; i < collection.size (); i++)
+	{
+	  stream << separator;
+	  visit (collection.at (i));
+	}
+    }
+}
+
+template <typename T>
+void
+Dump::visit_as_line (T &item, const std::string &trailing)
+{
+  stream << indentation;
+  visit (item);
+  stream << trailing << '\n';
+}
+
+template <typename T>
+void
+Dump::visit_items_as_lines (T &collection, const std::string &trailing)
+{
+  for (auto &item : collection)
+    visit_as_line (item);
```

Comment:
```suggestion
    visit_as_line (item, trailing);
```
Unless I'm misunderstanding something?